File: clients/android-discodrive/app/src/main/java/org/discodrive/android/autoupload/NameResolver.kt
package org.discodrive.android.autoupload

// Answers from the Go binding's Browser.existsWithHash.
const val EXISTS_ABSENT = "absent"
const val EXISTS_SAME = "same"
const val EXISTS_DIFFERENT = "different"

/**
 * Decides what name a file should land under in the destination folder.
 *
 * The server treats an upload with an existing name as a new version of that file, so a
 * photo named like one already there would quietly replace it. Every upload therefore asks
 * first, and a taken name gets a `-1`, `-2`, … suffix instead.
 */
object NameResolver {
    /** Beyond this something is wrong with the destination; deferring beats looping. */
    private const val MAX_TRIES = 50

    /**
     * Returns the name to upload under, or null when the file should be skipped — either
     * the identical bytes are already there, or no free name was found.
     *
     * [exists] takes a candidate name and answers [EXISTS_ABSENT] / [EXISTS_SAME] /
     * [EXISTS_DIFFERENT]; it is a lambda so this stays testable without the Go layer.
     */
    fun resolve(name: String, exists: (String) -> String): String? {
        for (attempt in 0..MAX_TRIES) {
            val candidate = if (attempt == 0) name else suffixed(name, attempt)
            when (exists(candidate)) {
                EXISTS_ABSENT -> return candidate
                EXISTS_SAME -> return null
                else -> Unit // taken by other content — try the next suffix
            }
        }
        return null
    }

    /** `IMG_1.jpg` + 2 → `IMG_1-2.jpg`; keeps dotfiles and multi-part extensions sane. */
    private fun suffixed(name: String, n: Int): String {
        val dot = name.lastIndexOf('.')
        // dot == 0 is a dotfile (".config"), not an extension.
        return if (dot <= 0) "$name-$n" else "${name.substring(0, dot)}-$n${name.substring(dot)}"
    }
}

File: clients/DiscoKit/Sources/DiscoKit/AutoUpload/NameResolver.swift
import Foundation

/// What sits at a name in the destination folder.
public enum NameState: Sendable, Equatable {
    case absent
    /// A file with the same content is already there.
    case same
    /// The name is taken by other content, or by a folder.
    case different
}

/// Decides what name a file should land under.
///
/// The server treats an upload with an existing name as a new version of that file, so a
/// photo named like one already there would quietly replace it. Every upload asks first,
/// and a taken name gets a `-1`, `-2`, … suffix instead.
public enum NameResolver {
    /// Beyond this something is wrong with the destination; deferring beats looping.
    static let maxTries = 50

    /// Returns the name to upload under, or nil when the file should be skipped — either
    /// the identical bytes are already there, or no free name was found.
    public static func resolve(_ name: String, exists: (String) -> NameState) -> String? {
        for attempt in 0...maxTries {
            let candidate = attempt == 0 ? name : suffixed(name, attempt)
            switch exists(candidate) {
            case .absent: return candidate
            case .same: return nil
            case .different: continue
            }
        }
        return nil
    }

    /// `IMG_1.jpg` + 2 → `IMG_1-2.jpg`; keeps dotfiles and multi-part extensions sane.
    static func suffixed(_ name: String, _ n: Int) -> String {
        guard let dot = name.lastIndex(of: "."), dot != name.startIndex else {
            return "\(name)-\(n)"
        }
        let base = name[name.startIndex..<dot]
        let ext = name[dot...]
        return "\(base)-\(n)\(ext)"
    }
}
